### PR TITLE
Assets for AI/ML Profiling session

### PR DESCRIPTION
**Pull request by khossain4337**
This pull request contains the PyTorch example, job submission scripts, and sample profiles for the AI/ML profiling session.

### CODE REVIEW

File: ai_ml_profiling/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+## Basic Strategy
+The main code allows us to play around with a few parameters, like the number of elements in the GPU buffer (the sequence length, the number that goes in the sequence dimension of the input to a transformer layer stack), the hidden dimension (a proxy for the model dimension), and the precision type (only `float32` and `bfloat16` is supported). We will try to record results for however many combinations we can go through. The lower cutoff of expectation here is to at least get profiles for 1 and 2 nodes, for both precision type. We will also try to pay attention to the results coming out of the timing loops.
+
+This will primarily be a two step process. Generate the profile on Polaris, `scp` or `rsync` it to the local machine (where we have `nsys` and `ncu` client/application/gui/viewer installed). Open the profiles and see if these make sense with respect to the workload that we have deployed.
+
+Installing Nsight Systems and Nsight Compute is fairly straightforward. We can start by doing a "download Nsight Systems" google search. In any case, the two links at the bottom should take us to the desired location!
+
+## Install Nsys
+
```
Comment (felker) [inline ai_ml_profiling/README.md:55]: Follow these links to install a version of NVIDIA's Nsight Systems and Nsight Compute on your local machine.

**Review by felker — CHANGES_REQUESTED** [review]
I agree with @nscottnichols comment about the binary profiles. You might want to move them to a public Box folder that you can share with attendees, in case they have issues generating the profiles.

File: ai_ml_profiling/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Profiling a simple PyTorch example
+In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. Very similar strategy is adopted by Megatron-DeepSpeed as well. 
+
+A request to an user would be to try and run the run the scripts both from an interactive session and a batch submission *before* jumping on to using the profilers!
```
[the remark this answers is not in the document]
Comment (khossain4337) [reply ai_ml_profiling/README.md:4]: It should be either! Fixing it!

File: ai_ml_profiling/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Profiling a simple PyTorch example
+In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. Very similar strategy is adopted by Megatron-DeepSpeed as well.
```
[the remark this answers is not in the document]
Comment (khossain4337) [reply ai_ml_profiling/README.md:2]: Adding this.

File: ai_ml_profiling/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Profiling a simple PyTorch example
+In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. A similar strategy is adopted by Megatron-DeepSpeed as well.
```
Comment (felker) [inline ai_ml_profiling/README.md:2]:
```suggestion
In the hands-on session we will try to profile a simple PyTorch example. This mini-application (or "mini-benchmark") mimics the sequence parallelism strategy that we use for our AuroraGPT training. A similar strategy is adopted by [Megatron-DeepSpeed](https://github.com/microsoft/Megatron-DeepSpeed) as well. 
```

File: ai_ml_profiling/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    #PBS -N the_name_you_like
+    #PBS -k doe
+    #PBS -o /path/to/the/stdout
+    #PBS -e /path/to/the/stderror
+
+    ...
+    WORK_DIR=/path/to/sequence_parallelism.py
+    PROF_DIR=/path/to/the/directory/for/profiles
+    ```
+    The `PROF_DIR` path is unused in this piece of code, as we have a separate script for launching the profiler.
```
Comment (felker) [inline ai_ml_profiling/README.md:29]: why not delete this line from the script, then?

File: ai_ml_profiling/qsub_profile_polaris_sequence_parallelism.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+let NRANKS=${NNODES}*${NRANKS_PER_NODE}
+
+module use /soft/modulefiles/
+module load conda/2024-04-29
+conda activate
+
+RUN_ID=profiling_seq_parallel_fp32-ranks${NRANKS}-nodes${NNODES}-T${TRIAL}
+
+mpiexec -n ${NRANKS} -ppn ${NRANKS_PER_NODE} --env TMPDIR=/eagle/datascience/hossainm/nsys_profile/tmpdir/ --cpu-bind=numa -l --line-buffer \
```
Comment (felker) [inline ai_ml_profiling/qsub_profile_polaris_sequence_parallelism.sh:35]: move TMPDIR to above, line after PROF_DIR, etc.

File: ai_ml_profiling/qsub_profile_polaris_sequence_parallelism.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#PBS -A datascience
+#PBS -N NSYS_1x2
+#PBS -k doe
+#PBS -o /home/hossainm/hpc_workshop_october_2024/outdir
+#PBS -e /home/hossainm/hpc_workshop_october_2024/errordir
+#PBS -j oe
+
+
+# What's the benchmark work directory?
+WORK_DIR=/home/hossainm/hpc_workshop_october_2024
```
Comment (felker) [inline ai_ml_profiling/qsub_profile_polaris_sequence_parallelism.sh:16]: I would suggest removing your username the RHS of the environment variables, and make it clearer that the user needs to fill that in. For example:

```suggestion
WORK_DIR=[INSERT PATH IN HOME DIR]
```
or
```suggestion
WORK_DIR=/home/[USERNAME]/hpc_workshop_october_2024
```


Same with the above PBS options with paths

File: ai_ml_profiling/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    Keeping track of the stdout (`.OU`) and the stderror (`.ER`) files are important, as we plan to ask the NVIDIA Nsight Systems profiler to print out a summary for us. This summary comes in the .OU file.
```
Comment (felker) [inline ai_ml_profiling/README.md:31]:
```suggestion
    Keeping track of the stdout (`.OU`) and the stderror (`.ER`) files are important, as we plan to ask the NVIDIA Nsight Systems profiler to print out a summary for us. This summary comes in the `.OU` file.
```

File: ai_ml_profiling/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# Profiling a simple PyTorch example
+In the hands on session we will try to profile a simple PyTorch example. This mini-application (mini-benchmark) mimics the sequence parallelism strategy that we use for our auroraGPT training. A similar strategy is adopted by Megatron-DeepSpeed as well. 
+
+A request to an user would be to try and run the run the scripts either from an interactive session or a batch submission *before* jumping on to using the profilers!
```
Comment (felker) [inline ai_ml_profiling/README.md:4]:
```suggestion
Users should first attempt to run the following workload scripts without profiling on a Polaris compute node (from either an interactive session or a batch job submission).
```

File: ai_ml_profiling/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+This will primarily be a two step process. Generate the profile on Polaris, `scp` or `rsync` it to the local machine (where we have `nsys` and `ncu` client/application/gui/viewer installed). Open the profiles and see if these make sense with respect to the workload that we have deployed.
+
+Follow the links below to install a version of NVIDIA's Nsight Systems and Nsight Compute on your local machine.
+
+## Install Nsys
+
+[Getting Started, Download Nsys](https://developer.nvidia.com/nsight-systems/get-started)
+
+[Download Nsight Compute](https://developer.nvidia.com/tools-overview/nsight-compute/get-started)
```
Comment (felker) [inline ai_ml_profiling/README.md:58]: Would be nice to add a short section about `## View profiles` with some guiding hints on what to look for, what parameters you modify (unless you have that content planned for slides, etc.) Could also add a link to a public Box folder containing the `qdrep` etc. binary profile data for folks who are unable to collect the performance profiles themselves, for whatever reason.

**Review by felker — CHANGES_REQUESTED** [review]

**Pull request merged**